astroid/brain/brain_dataclasses.py

```diff
@@ -190,12 +195,12 @@ def _generate_dataclass_init(assigns: List[AnnAssign]) -> str:
             if isinstance(value, Call) and _looks_like_dataclass_field_call(
                 value, check_scope=False
             ):
-                result = _get_field_default(value)
-
-                default_type, default_node = result
+                default_type, default_node = _get_field_default(value)
```

Comment:
@cdce8p For some reason I can't seem to restrict the return type of `_get_field_default` and let `mypy` know about this.

I tried `Union[Tuple[None, None], Tuple[str, NodeNG]]` (and the necessary change to the method) as return type annotation, but then `mypy` doesn't recognise that `if default_type == "default"` then `default_node` is a `NodeNG`. I have tried other things such as working but a type alias but I couldn't get it to work. Searching for "mypy union return distinguish tuple constraints" etc also didn't return anything.

Do you have any good ideas or are these `assert`'s the best solution here?

Reply:
Here's a neat way to do that:

```python
result = _get_field_default(value)
if result != (None, None):
    default_type, default_node = result
    # Here mypy should know that default_type is a str, and default_node is a NodeNG
```

Reply:
You could also modify the return type of `_get_field_value` to return `None` instead of `None, None`. Might make it better?

Reply:
```python
                result = _get_field_default(value)
                if result != (None, None):
                    default_type, default_node = result
                    if default_type == "default":
                        param_str += f" = {default_node.as_string()}"  # L202
                    elif default_type == "default_factory":
                        param_str += f" = {DEFAULT_FACTORY}"
                        assignment_str = (
                            f"self.{name} = {default_node.as_string()} "  # L206
                            f"if {name} is {DEFAULT_FACTORY} else {name}"
                        )
```

Still gives me:
```shell
astroid/brain/brain_dataclasses.py:202: error: Item "None" of "Optional[NodeNG]" has no attribute "as_string"  [union-attr]
astroid/brain/brain_dataclasses.py:206: error: Item "None" of "Optional[NodeNG]" has no attribute "as_string"  [union-attr]
```

Am I missing something?

Reply:
Did you change the return type for that function to `Union[Tuple[None, None], Tuple[str, NodeNG]]`?

Reply:
Yeah.

Diff:
<details>
diff --git a/astroid/brain/brain_dataclasses.py b/astroid/brain/brain_dataclasses.py
index b34f1aeff..d239806c5 100644
--- a/astroid/brain/brain_dataclasses.py
+++ b/astroid/brain/brain_dataclasses.py
@@ -7,7 +7,7 @@ Support both built-in dataclasses and pydantic.dataclasses. References:
 - https://docs.python.org/3/library/dataclasses.html
 - https://pydantic-docs.helpmanual.io/usage/dataclasses/
 """
-from typing import FrozenSet, Generator, List, Optional, Tuple
+from typing import FrozenSet, Generator, List, Optional, Tuple, Union
 
 from astroid import context, inference_tip
 from astroid.builder import parse
@@ -195,17 +195,18 @@ def _generate_dataclass_init(assigns: List[AnnAssign]) -> str:
             if isinstance(value, Call) and _looks_like_dataclass_field_call(
                 value, check_scope=False
             ):
-                default_type, default_node = _get_field_default(value)
-                if default_type == "default":
-                    assert default_node
-                    param_str += f" = {default_node.as_string()}"
-                elif default_type == "default_factory":
-                    assert default_node
-                    param_str += f" = {DEFAULT_FACTORY}"
-                    assignment_str = (
-                        f"self.{name} = {default_node.as_string()} "
-                        f"if {name} is {DEFAULT_FACTORY} else {name}"
-                    )
+                result = _get_field_default(value)
+                if result != (None, None):
+                    default_type, default_node = result
+                    if default_type == "default":
+                        param_str += f" = {default_node.as_string()}"
+                    elif default_type == "default_factory":
+                        param_str += f" = {DEFAULT_FACTORY}"
+                        assignment_str = (
+                            f"self.{name} = {default_node.as_string()} "
+                            f"if {name} is {DEFAULT_FACTORY} else {name}"
+                        )
+
             else:
                 param_str += f" = {value.as_string()}"
 
@@ -335,7 +336,9 @@ def _looks_like_dataclass_field_call(node: Call, check_scope: bool = True) -> bo
     return inferred.name == FIELD_NAME and inferred.root().name in DATACLASS_MODULES
 
 
-def _get_field_default(field_call: Call) -> Tuple[str, Optional[NodeNG]]:
+def _get_field_default(
+    field_call: Call,
+) -> Union[Tuple[None, None], Tuple[str, NodeNG]]:
     """Return a the default value of a field call, and the corresponding keyword argument name.
 
     field(default=...) results in the ... node
@@ -363,7 +366,7 @@ def _get_field_default(field_call: Call) -> Tuple[str, Optional[NodeNG]]:
         new_call.postinit(func=default_factory)
         return "default_factory", new_call
 
-    return "", None
+    return None, None
 
 
 def _is_class_var(node: Optional[NodeNG]) -> bool:
</details>

Reply:
The following diff works, but it feels weird to need to do an additional assignment to help `mypy` infer the type:
It should be able to link the `None` to the `None` and `str` to `NodeNG`.

```diff
diff --git a/astroid/brain/brain_dataclasses.py b/astroid/brain/brain_dataclasses.py
index b34f1aeff..648fdb827 100644
--- a/astroid/brain/brain_dataclasses.py
+++ b/astroid/brain/brain_dataclasses.py
@@ -7,7 +7,7 @@ Support both built-in dataclasses and pydantic.dataclasses. References:
 - https://docs.python.org/3/library/dataclasses.html
 - https://pydantic-docs.helpmanual.io/usage/dataclasses/
 """
-from typing import FrozenSet, Generator, List, Optional, Tuple
+from typing import FrozenSet, Generator, List, Optional, Tuple, Union
 
 from astroid import context, inference_tip
 from astroid.builder import parse
@@ -195,17 +195,17 @@ def _generate_dataclass_init(assigns: List[AnnAssign]) -> str:
             if isinstance(value, Call) and _looks_like_dataclass_field_call(
                 value, check_scope=False
             ):
-                default_type, default_node = _get_field_default(value)
-                if default_type == "default":
-                    assert default_node
-                    param_str += f" = {default_node.as_string()}"
-                elif default_type == "default_factory":
-                    assert default_node
-                    param_str += f" = {DEFAULT_FACTORY}"
-                    assignment_str = (
-                        f"self.{name} = {default_node.as_string()} "
-                        f"if {name} is {DEFAULT_FACTORY} else {name}"
-                    )
+                result = _get_field_default(value)
+                if result:
+                    default_type, default_node = result
+                    if default_type == "default":
+                        param_str += f" = {default_node.as_string()}"
+                    elif default_type == "default_factory":
+                        param_str += f" = {DEFAULT_FACTORY}"
+                        assignment_str = (
+                            f"self.{name} = {default_node.as_string()} "
+                            f"if {name} is {DEFAULT_FACTORY} else {name}"
+                        )
             else:
                 param_str += f" = {value.as_string()}"
 
@@ -335,7 +335,9 @@ def _looks_like_dataclass_field_call(node: Call, check_scope: bool = True) -> bo
     return inferred.name == FIELD_NAME and inferred.root().name in DATACLASS_MODULES
 
 
-def _get_field_default(field_call: Call) -> Tuple[str, Optional[NodeNG]]:
+def _get_field_default(
+    field_call: Call,
+) -> Optional[Tuple[str, NodeNG]]:
     """Return a the default value of a field call, and the corresponding keyword argument name.
 
     field(default=...) results in the ... node
@@ -363,7 +365,7 @@ def _get_field_default(field_call: Call) -> Tuple[str, Optional[NodeNG]]:
         new_call.postinit(func=default_factory)
         return "default_factory", new_call
 
-    return "", None
+    return None
 
 
 def _is_class_var(node: Optional[NodeNG]) -> bool:
```

Reply:
Yeah. Possibly mypy bug.

But I prefer this solution, as we were essentially returning two sentinels in the previous case.